runtime/src/server.ts

```diff
@@ -36,20 +38,31 @@ export interface EndpointArg<
   query: Q;
   body: Body;
 }
+
+type ServerEndpointArg<
+  H extends Headers | void,
+  P extends Params | void,
+  Q extends Query | void,
+  Body extends RequestBody<any> | void,
+  RC extends RequestContext
+> = EndpointArg<H, P, Q, Body> & { readonly requestContext: RC };
+
 export type Endpoint<
   H extends Headers | void,
   P extends Params | void,
   Q extends Query | void,
   Body extends RequestBody<any> | void,
-  R extends Response<number, any, any>
-> = (ctx: EndpointArg<H, P, Q, Body>) => Promise<R>;
+  R extends Response<number, any, any>,
+  RC extends RequestContext
+> = (ctx: ServerEndpointArg<H, P, Q, Body, RC>) => Promise<R>;
 
 export type SafeEndpoint = Endpoint<
   Headers | undefined,
   Params | undefined,
   Query | undefined,
   RequestBody<any> | undefined,
-  Response<number, any, any>
+  Response<number, any, any>,
+  RequestContext
```

Comment:
Why don't you incapsulate `Headers`, `Params`, and `Query` in the `RequestContext`?

Reply:
Because they in no way belong there. RequestContext can be used to provide request specific facilities to endpoint handlers, like logging & authorization. As an example, you can check Caesar PR https://github.com/smartlyio/caesar/pull/180 (work in progress).